src/Kerbalism/System/Kerbalism.cs

```diff
@@ -490,6 +392,49 @@ void FixedUpdate()
 			fuWatch.Stop();
 		}
 
+		public void ManageWorkQueues(Queue<Vessel> mandatory, Queue<Vessel> discretionary)
```

Comment:
What's not simple about a Queue?  Because you have a queue: an collection of things to process that you have no need to revisit or index into arbitrarily, and for which adding and removing are extremely fast.

The point of doing this was somewhat that the oldest vessel is irrelevant, and *if you ever want to do more than 1*, what is the next-oldest vessel?  It doesn't matter.  It's not one you've processed since you filled the queue.  The background queue doesn't need to refresh every FixedUpdate, you don't need to find the oldest vessel.  The oldest vessel is always in the queue, because the only way out of the queue is to get processed.  In fact, everything in the queue is basically of similar-enough age, and if the global list of vessels itself doesn't constantly re-order, the natural ordering of creating the queue will also be oldest-first after the first complete run... so you get "oldest vessel" free *anyway*.  In fact, you get a sorted queue for free after the first cycle.

I think we both accuse eachother a lot of premature optimization.  Dictionary lookup is O(1).  Extra TryGetVesselData is pretty insignificant.  I do keep looking where I can iterate over the VesselDatas instead of Vessels because that's usually what I really want, anyway.  Here was a case where I thought not to.  Because:

Access to FlightGlobals' short list of loaded vessels are the only ones considered this cycle.  I don't even need to inspect all of the unloaded ones.  Why bother to iterate over a large collection when I can iterate over a tiny one and do fast Dictionary lookups to double-check validity?  And I'm actually doing relatively few of them.  Unloaded vessels only need to be reviewed every complete block -- they can be checked when the queue refills.

When you complete a full pass on all vessels, you get a reasonably-sized chunk of expired frames to return.
When you complete a single vessel, you get a small number of frames to return. (Exactly 1 FU's generation, I believe.)  And it is the same cost to find them all each time.

Why 2 queues, beyond that they can have different update rates?  Because now, with a separate queue of background vessels to process as you get to them, minimum 1/cycle, you could dynamically decide whether to pull from that queue or not.  Consider: start a stopwatch, process a vessel, check elapsed time, if within budget, grab the next vessel and process.  Which vessel?  The next one in the queue.

I think the correct implementation is, well, what I wrote.

Reply:
Uh, I didn't understand that the "discretionary" queue was long lived.
Well, we need to iterate on all vessels in any case (ie, `ValidateVesselData()` **must** be called every fixedupdate).
Said otherwise, it is mandatory that the DB.VesselData and FlightGlobals.Vessels are synchronized before e do anything. So we have to iterate on both collections every fixedUpdate in any case. So there is little point in caching anything.

Reply:
I don't understand.
ValidateVesselData creates a VesselData object in the DB if one doesn't exist.

I can understand catching any loaded vessels that didn't get a DB entry here, if we don't trust the other places in the code to detect new loaded vessels and set them up.  I didn't include a fallback, but easy enough to (since I already check that a valid VesselData exists)

For background vessels?  What exactly are we hunting?  If it went (NumVessels) FixedUpdates cycles before being detected, does that matter?  What's going to process it if it doesn't have a VesselData?  There are things that intentionally won't, since there is: `if (VesselData.VesselNeedVesselData(v.protoVessel))` in there.

If it is necessary, perhaps move it up out of the substep sim and the sim's decision-making on what to definitely process and what to maybe process this cycle?

Reply:
There are tons of entry points for getting unloaded vessels VesselData (CommNet, GameEvents...)
Anyway, the main issue is that long lived queue. There are tons of ways the discretionary queue will become invalid : vessel changing loaded state, vessel changing simulated state, destroyed vessels... It's an unnecessary point of failure.

Reply:
I've moved ValidateVesselData to a point where it will always run.

I have also swapped the queues to more directly reflect loaded and unloaded, instead of mandatory/discretionary, as part of 47533e8.  That allows flexibility in how we choose the next vessel to process while we wait for jobs to finish.  As an idea, there is a configurable time budget to process unloaded vessels from before switching to loaded.  The purpose is to provide useful main-thread processing while we wait for potentially-expensive jobs to complete.  (40 vessels 100k timewarp typical 3ms and without access to the main thread, so "expensive" is relative.)